// @react-pdf-viewer/full-screen/lib/cjs/full-screen.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var React = require('react');
var core = require('@react-pdf-viewer/core');

function _interopNamespace(e) {
    if (e && e.__esModule) return e;
    var n = Object.create(null);
    if (e) {
        Object.keys(e).forEach(function (k) {
            if (k !== 'default') {
                var d = Object.getOwnPropertyDescriptor(e, k);
                Object.defineProperty(n, k, d.get ? d : {
                    enumerable: true,
                    get: function () {
                        return e[k];
                    }
                });
            }
        });
    }
    n['default'] = e;
    return Object.freeze(n);
}

var React__namespace = /*#__PURE__*/_interopNamespace(React);

var ExitFullScreenIcon = function () {
    return (React__namespace.createElement(core.Icon, { size: 16 },
        React__namespace.createElement("path", { d: "M11.5 23.499L11.5 14.499" }),
        React__namespace.createElement("path", { d: "M7.5 18.499L11.5 14.499 15.5 18.499" }),
        React__namespace.createElement("path", { d: "M11.5 1.499L11.5 10.499" }),
        React__namespace.createElement("path", { d: "M7.5 6.499L11.5 10.499 15.5 6.499" }),
        React__namespace.createElement("path", { d: "M20.5 12.499L1.5 12.499" })));
};

/*! *****************************************************************************
Copyright (c) Microsoft Corporation.

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.
***************************************************************************** */

var __assign = function() {
    __assign = Object.assign || function __assign(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};

var FullScreenIcon = function () {
    return (React__namespace.createElement(core.Icon, { size: 16 },
        React__namespace.createElement("path", { d: "M0.5 12L23.5 12" }),
        React__namespace.createElement("path", { d: "M11.5 1L11.5 23" }),
        React__namespace.createElement("path", { d: "M8.5 4L11.5 1 14.5 4" }),
        React__namespace.createElement("path", { d: "M20.5 9L23.5 12 20.5 15" }),
        React__namespace.createElement("path", { d: "M3.5 15L0.5 12 3.5 9" }),
        React__namespace.createElement("path", { d: "M14.5 20L11.5 23 8.5 20" })));
};

var TOOLTIP_OFFSET = { left: 0, top: 8 };
var EnterFullScreenButton = function (_a) {
    var onClick = _a.onClick;
    var l10n = React__namespace.useContext(core.LocalizationContext);
    var label = l10n && l10n.fullScreen ? l10n.fullScreen.enterFullScreen : 'Full screen';
    return (React__namespace.createElement(core.Tooltip, { position: core.Position.BottomCenter, target: React__namespace.createElement(core.MinimalButton, { onClick: onClick },
            React__namespace.createElement(FullScreenIcon, null)), content: function () { return label; }, offset: TOOLTIP_OFFSET }));
};

var Api;
(function (Api) {
    Api[Api["ExitFullScreen"] = 0] = "ExitFullScreen";
    Api[Api["FullScreenChange"] = 1] = "FullScreenChange";
    Api[Api["FullScreenElement"] = 2] = "FullScreenElement";
    Api[Api["FullScreenEnabled"] = 3] = "FullScreenEnabled";
    Api[Api["RequestFullScreen"] = 4] = "RequestFullScreen";
})(Api || (Api = {}));
var defaultVendor = {
    ExitFullScreen: 'exitFullscreen',
    FullScreenChange: 'fullscreenchange',
    FullScreenElement: 'fullscreenElement',
    FullScreenEnabled: 'fullscreenEnabled',
    RequestFullScreen: 'requestFullscreen',
};
var webkitVendor = {
    ExitFullScreen: 'webkitExitFullscreen',
    FullScreenChange: 'webkitfullscreenchange',
    FullScreenElement: 'webkitFullscreenElement',
    FullScreenEnabled: 'webkitFullscreenEnabled',
    RequestFullScreen: 'webkitRequestFullscreen',
};
var msVendor = {
    ExitFullScreen: 'msExitFullscreen',
    FullScreenChange: 'MSFullscreenChange',
    FullScreenElement: 'msFullscreenElement',
    FullScreenEnabled: 'msFullscreenEnabled',
    RequestFullScreen: 'msRequestFullscreen',
};
var isBrowser = typeof window !== 'undefined';
var vendor = isBrowser
    ? (Api.FullScreenEnabled in document && defaultVendor) ||
        (webkitVendor.FullScreenEnabled in document && webkitVendor) ||
        (msVendor.FullScreenEnabled in document && msVendor) ||
        defaultVendor
    : defaultVendor;
var addFullScreenChangeListener = function (handler) {
    if (isBrowser) {
        document.addEventListener(vendor.FullScreenChange, handler);
    }
};
var exitFullScreen = function (element) {
    return isBrowser
        ?
            element[vendor.ExitFullScreen]()
        : Promise.resolve({});
};
var getFullScreenElement = function () {
    return isBrowser ? document[vendor.FullScreenElement] : null;
};
var requestFullScreen = function (element) {
    if (isBrowser) {
        element[vendor.RequestFullScreen]();
    }
};

var EnterFullScreen = function (_a) {
    var children = _a.children, store = _a.store, onEnterFullScreen = _a.onEnterFullScreen, onExitFullScreen = _a.onExitFullScreen;
    var pagesRef = React__namespace.useRef(store.get('getPagesContainer') ? store.get('getPagesContainer')() : null);
    var closeOtherFullScreen = function () {
        var pagesEle = pagesRef.current;
        if (!pagesEle) {
            return Promise.resolve();
        }
        var ele = getFullScreenElement();
        return ele && ele !== pagesEle ? exitFullScreen(ele) : Promise.resolve();
    };
    var enterFullScreen = function () {
        var pagesEle = pagesRef.current;
        if (!pagesEle) {
            return;
        }
        closeOtherFullScreen().then(function () {
            requestFullScreen(pagesEle);
        });
    };
    var onFullScreenChange = function () {
        var ele = getFullScreenElement();
        var isFullScreen = ele === pagesRef.current;
        store.update('isFullScreen', isFullScreen);
        var zoom = store.get('zoom');
        if (zoom) {
            isFullScreen ? onEnterFullScreen(zoom) : onExitFullScreen(zoom);
        }
    };
    var handlePagesContainer = function (getPagesContainer) {
        pagesRef.current = getPagesContainer();
        addFullScreenChangeListener(onFullScreenChange);
    };
    React__namespace.useEffect(function () {
        store.subscribe('getPagesContainer', handlePagesContainer);
        return function () {
            store.unsubscribe('getPagesContainer', handlePagesContainer);
        };
    }, []);
    var defaultChildren = function (props) { return React__namespace.createElement(EnterFullScreenButton, { onClick: props.onClick }); };
    var render = children || defaultChildren;
    return render({
        onClick: enterFullScreen,
    });
};

var EnterFullScreenMenuItem = function (_a) {
    var onClick = _a.onClick;
    var l10n = React__namespace.useContext(core.LocalizationContext);
    var label = l10n && l10n.fullScreen ? l10n.fullScreen.enterFullScreen : 'Full screen';
    return (React__namespace.createElement(core.MenuItem, { icon: React__namespace.createElement(FullScreenIcon, null), onClick: onClick }, label));
};

var ExitFullScreenButton = function (_a) {
    var store = _a.store;
    var _b = React__namespace.useState(false), isFullScreen = _b[0], setFullScreen = _b[1];
    var handleFullScreen = function (fullScreen) {
        setFullScreen(fullScreen);
    };
    var handleExitFullScreen = function () {
        setFullScreen(false);
        var getPagesContainer = store.get('getPagesContainer');
        if (!getPagesContainer) {
            return;
        }
        var pagesEle = getPagesContainer();
        if (!pagesEle) {
            return;
        }
        var ele = getFullScreenElement();
        if (ele && ele === pagesEle) {
            exitFullScreen(document);
        }
    };
    React__namespace.useEffect(function () {
        store.subscribe('isFullScreen', handleFullScreen);
        return function () {
            store.unsubscribe('isFullScreen', handleFullScreen);
        };
    }, []);
    return (React__namespace.createElement(React__namespace.Fragment, null, isFullScreen && (React__namespace.createElement("div", { className: "rpv-full-screen__exit-button" },
        React__namespace.createElement("div", { className: "rpv-full-screen__exit-button-content" },
            React__namespace.createElement(core.MinimalButton, { onClick: handleExitFullScreen },
                React__namespace.createElement(ExitFullScreenIcon, null)))))));
};

var fullScreenPlugin = function (props) {
    var store = React__namespace.useMemo(function () { return core.createStore({}); }, []);
    var onEnterFullScreen = props && props.onEnterFullScreen ? props.onEnterFullScreen : function () { };
    var onExitFullScreen = props && props.onExitFullScreen ? props.onExitFullScreen : function () { };
    var EnterFullScreenDecorator = function (props) { return (React__namespace.createElement(EnterFullScreen, __assign({}, props, { store: store, onEnterFullScreen: onEnterFullScreen, onExitFullScreen: onExitFullScreen }))); };
    var EnterFullScreenButtonDecorator = function () { return (React__namespace.createElement(EnterFullScreenDecorator, null, function (renderProps) { return React__namespace.createElement(EnterFullScreenButton, __assign({}, renderProps)); })); };
    var EnterFullScreenMenuItemDecorator = function (props) { return (React__namespace.createElement(EnterFullScreenDecorator, null, function (p) { return (React__namespace.createElement(EnterFullScreenMenuItem, { onClick: function () {
            p.onClick();
            props.onClick();
        } })); })); };
    var ExitFullScreenDecorator = function () { return React__namespace.createElement(ExitFullScreenButton, { store: store }); };
    var renderViewer = function (props) {
        var currentSlot = props.slot;
        if (currentSlot.subSlot) {
            currentSlot.subSlot.children = (React__namespace.createElement(React__namespace.Fragment, null,
                React__namespace.createElement(ExitFullScreenDecorator, null),
                currentSlot.subSlot.children));
        }
        return currentSlot;
    };
    return {
        install: function (pluginFunctions) {
            store.update('getPagesContainer', pluginFunctions.getPagesContainer);
            store.update('zoom', pluginFunctions.zoom);
        },
        renderViewer: renderViewer,
        EnterFullScreen: EnterFullScreenDecorator,
        EnterFullScreenButton: EnterFullScreenButtonDecorator,
        EnterFullScreenMenuItem: EnterFullScreenMenuItemDecorator,
    };
};

exports.ExitFullScreenIcon = ExitFullScreenIcon;
exports.FullScreenIcon = FullScreenIcon;
exports.fullScreenPlugin = fullScreenPlugin;
